File: src/mitchell/store.py
"""MitchellStore, Event, and Analysis — core public API."""

from __future__ import annotations

import os
from pathlib import Path
from typing import TYPE_CHECKING, Iterator

import numpy as np
import zarr

if TYPE_CHECKING:
    import arviz as az

# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

_META_ATTRS = [
    "approximant", "cosmology", "delta_f", "distance_marginalization",
    "duration", "f_final", "f_low", "f_ref", "phase_marginalization",
    "reference_frame", "sampling_frequency", "start_time",
    "time_marginalization", "time_reference",
]


def _decode(val: object) -> object:
    """Unwrap length-1 arrays and decode byte strings for zarr attributes."""
    if isinstance(val, np.ndarray) and val.ndim == 1 and len(val) == 1:
        val = val[0]
    if isinstance(val, (bytes, np.bytes_)):
        val = val.decode()
    return val


def _columnar_to_dict(group: zarr.Group) -> dict[str, np.ndarray]:
    """Return {param: array} for every array dataset in *group*."""
    return {k: group[k][:] for k in group.array_keys()}


# ---------------------------------------------------------------------------
# Analysis
# ---------------------------------------------------------------------------

class Analysis:
    """A single PE analysis (waveform approximant run) within an event.

    Parameters
    ----------
    group:
        The zarr group at ``events/{event_name}/{analysis_name}``.
    """

    def __init__(self, group: zarr.Group) -> None:
        self._group = group

    @property
    def name(self) -> str:
        return self._group.name.split("/")[-1]

    @property
    def attrs(self) -> zarr.attrs.Attributes:
        return self._group.attrs

    # -- data product properties --------------------------------------------

    @property
    def posterior(self) -> zarr.Group | None:
        """Zarr group of columnar posterior sample arrays, or None."""
        if "posterior" in self._group and "samples" in self._group["posterior"]:
            return self._group["posterior"]["samples"]
        return None

    @property
    def priors(self) -> zarr.Group | None:
        """Zarr group containing ``samples`` and/or ``analytic`` sub-groups."""
        if "priors" in self._group:
            return self._group["priors"]
        return None

    @property
    def psds(self) -> zarr.Group | None:
        """Zarr group of PSD arrays keyed by detector name."""
        if "psds" in self._group:
            return self._group["psds"]
        return None

    @property
    def calibration_envelope(self) -> zarr.Group | None:
        """Zarr group of calibration envelope arrays keyed by detector."""
        if "calibration_envelope" in self._group:
            return self._group["calibration_envelope"]
        return None

    @property
    def skymap(self) -> zarr.Group | None:
        """Zarr group containing the HEALPix skymap ``data`` array."""
        if "skymap" in self._group:
            return self._group["skymap"]
        return None

    @property
    def config(self) -> dict[str, object]:
        """Config file key-value pairs as a plain dict."""
        if "config_file" in self._group:
            return dict(self._group["config_file"].attrs)
        return {}

    # -- ArviZ conversion ---------------------------------------------------

    def to_arviz(self) -> az.InferenceData:
        """Convert to an :class:`arviz.InferenceData` object.

        Posterior and prior samples are each loaded into memory and placed
        into the corresponding ArviZ groups.  All other data products
        (PSDs, calibration, skymap) are attached as
        ``constant_data`` / ``observed_data`` where appropriate.

        Returns
        -------
        arviz.InferenceData
        """
        import arviz as az

        data: dict[str, dict[str, np.ndarray]] = {}

        if self.posterior is not None:
            post = _columnar_to_dict(self.posterior)
            # ArviZ expects (chain, draw, ...) — reshape flat draws to (1, N)
            data["posterior"] = {k: v[np.newaxis, :] for k, v in post.items()}

        if self.priors is not None and "samples" in self.priors:
            prior = _columnar_to_dict(self.priors["samples"])
            data["prior"] = {k: v[np.newaxis, :] for k, v in prior.items()}

        if self.psds is not None:
            psd_data = {det: self.psds[det][:] for det in self.psds.array_keys()}
            data["observed_data"] = psd_data

        return az.from_dict(data)

    def __repr__(self) -> str:
        return f"Analysis({self.name!r})"


# ---------------------------------------------------------------------------
# Event
# ---------------------------------------------------------------------------

class Event:
    """A gravitational-wave event containing one or more analyses.

    Parameters
    ----------
    group:
        The zarr group at ``events/{event_name}``.
    """

    def __init__(self, group: zarr.Group) -> None:
        self._group = group

    @property
    def name(self) -> str:
        return self._group.name.split("/")[-1]

    def __getitem__(self, analysis_name: str) -> Analysis:
        return Analysis(self._group[analysis_name])

    def __iter__(self) -> Iterator[Analysis]:
        for key in self._group.group_keys():
            yield Analysis(self._group[key])

    def __len__(self) -> int:
        return sum(1 for _ in self._group.group_keys())

    def __repr__(self) -> str:
        names = [a.name for a in self]
        return f"Event({self.name!r}, analyses={names})"


# ---------------------------------------------------------------------------
# MitchellStore
# ---------------------------------------------------------------------------

class MitchellStore:
    """A zarr-backed store of gravitational-wave analysis data products.

    Open an existing store or create a new one with
    :meth:`MitchellStore.open`.  To import from a GWTC-style bilby HDF5
    file use :meth:`MitchellStore.from_bilby_h5`.

    Parameters
    ----------
    store:
        Any zarr v3 :class:`zarr.storage.StoreLike`.
    mode:
        Passed to :func:`zarr.open_group`.  Use ``"r"`` for read-only,
        ``"a"`` to open existing or create, ``"w"`` to overwrite.
    """

    SCHEMA_VERSION = "0.1.0"

    def __init__(
        self,
        store: zarr.storage.StoreLike,
        mode: str = "a",
    ) -> None:
        self._store = store
        self._root = zarr.open_group(store=store, mode=mode)
        if mode != "r" and "mitchell_schema_version" not in self._root.attrs:
            self._root.attrs["mitchell_schema_version"] = self.SCHEMA_VERSION

    # -- constructors -------------------------------------------------------

    @classmethod
    def open(
        cls,
        path: str | os.PathLike,
        mode: str = "a",
    ) -> MitchellStore:
        """Open (or create) a store at *path* on the local filesystem."""
        store = zarr.storage.LocalStore(Path(path))
        return cls(store, mode=mode)

    @classmethod
    def from_bilby_h5(
        cls,
        h5_path: str | os.PathLike,
        event_name: str,
        zarr_path: str | os.PathLike,
    ) -> MitchellStore:
        """Translate a GWTC bilby HDF5 file into a new Mitchell zarr store.

        Parameters
        ----------
        h5_path:
            Path to the source ``*.h5`` file.
        event_name:
            Human-readable event identifier, e.g. ``"GW150914_095045"``.
        zarr_path:
            Destination directory for the new zarr store.

        Returns
        -------
        MitchellStore
            The newly created store, opened in append mode.
        """
        import h5py as h5

        instance = cls.open(zarr_path, mode="w")
        root = instance._root
        event = root.require_group(f"events/{event_name}")

        with h5.File(h5_path, "r") as f:
            for analysis_name, h5_group in f.items():
                analysis = event.require_group(analysis_name)

                # -- meta_data attributes --
                if (
                    "meta_data" in h5_group
                    and "meta_data" in h5_group["meta_data"]
                ):
                    md = h5_group["meta_data"]["meta_data"]
                    analysis.attrs.update(
                        {k: _decode(md[k][()]) for k in _META_ATTRS if k in md}
                    )

                if "version" in h5_group:
                    analysis.attrs["version"] = _decode(h5_group["version"][()])
                if "description" in h5_group:
                    analysis.attrs["description"] = _decode(
                        h5_group["description"][()]
                    )

                # -- posterior samples --
                if "posterior_samples" in h5_group:
                    samples = h5_group["posterior_samples"][()]
                    post_group = analysis.require_group("posterior/samples")
                    for param in samples.dtype.names:
                        post_group[param] = samples[param]

                # -- prior samples + analytic priors --
                if "priors" in h5_group:
                    priors = h5_group["priors"]
                    if "samples" in priors:
                        prior_samples = analysis.require_group("priors/samples")
                        for param, dataset in priors["samples"].items():
                            prior_samples[param] = dataset[()]
                    if "analytic" in priors and len(priors["analytic"]) > 0:
                        analytic_group = analysis.require_group("priors/analytic")
                        analytic_group.attrs.update(
                            {
                                k: _decode(v[()])
                                for k, v in priors["analytic"].items()
                                if isinstance(v, h5.Dataset)
                            }
                        )

                # -- PSDs --
                if "psds" in h5_group:
                    psds_group = analysis.require_group("psds")
                    for detector, dataset in h5_group["psds"].items():
                        psds_group[detector] = dataset[()]

                # -- calibration envelope --
                if "calibration_envelope" in h5_group:
                    cal_group = analysis.require_group("calibration_envelope")
                    for detector, dataset in h5_group[
                        "calibration_envelope"
                    ].items():
                        cal_group[detector] = dataset[()]

                # -- config file --
                if (
                    "config_file" in h5_group
                    and "config" in h5_group["config_file"]
                ):
                    config_group = analysis.require_group("config_file")
                    config_group.attrs.update(
                        {
                            k: _decode(v[()])
                            for k, v in h5_group["config_file"][
                                "config"
                            ].items()
                            if isinstance(v, h5.Dataset)
                        }
                    )

                # -- skymap --
                if "skymap" in h5_group:
                    skymap_group = analysis.require_group("skymap")
                    skymap_h5 = h5_group["skymap"]
                    if "data" in skymap_h5:
                        skymap_group["data"] = skymap_h5["data"][()]
                    if "meta_data" in skymap_h5:
                        skymap_group.attrs.update(
                            {
                                k: _decode(v[()])
                                for k, v in skymap_h5["meta_data"].items()
                                if isinstance(v, h5.Dataset)
                            }
                        )

        zarr.consolidate_metadata(instance._store)
        return instance

    # -- event access -------------------------------------------------------

    def __getitem__(self, event_name: str) -> Event:
        return Event(self._root["events"][event_name])

    def __iter__(self) -> Iterator[Event]:
        if "events" not in self._root:
            return
        for key in self._root["events"].group_keys():
            yield Event(self._root["events"][key])

    def __len__(self) -> int:
        if "events" not in self._root:
            return 0
        return sum(1 for _ in self._root["events"].group_keys())

    def __repr__(self) -> str:
        n = len(self)
        return f"MitchellStore({n} event{'s' if n != 1 else ''})"

File: bindings/javascript/src/store.ts
/**
 * Core Mitchell API: {@link MitchellStore}, {@link Event}, and {@link Analysis}.
 *
 * The three classes mirror the Python ``mitchell`` package.  All
 * discovery (listing events, analyses, parameter names) is performed
 * synchronously from the consolidated metadata embedded in the root
 * ``zarr.json``; only actual array data loads are asynchronous.
 *
 * @example
 * ```typescript
 * import { MitchellStore } from "@mitchell/javascript";
 * import { FileSystemStore } from "@mitchell/javascript/node";
 *
 * const ms = await MitchellStore.open(new FileSystemStore("events.zarr"));
 * for await (const event of ms.events()) {
 *   for await (const analysis of event.analyses()) {
 *     const mass1 = await analysis.posterior!.get("mass_1");
 *   }
 * }
 * ```
 *
 * @module
 */

import * as zarrita from "zarrita";
import type { Readable } from "@zarrita/storage";
import type {
  AnalysisAttrs,
  ConsolidatedEntry,
  MitchellRootMetadata,
} from "./types.js";

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

const decoder = new TextDecoder();

/** Read the root zarr.json and return the parsed object. */
async function readRootJson(store: Readable): Promise<MitchellRootMetadata> {
  const bytes = await store.get("/zarr.json");
  if (!bytes) {
    throw new Error("Not a valid zarr store: /zarr.json not found");
  }
  return JSON.parse(decoder.decode(bytes)) as MitchellRootMetadata;
}

/**
 * Return names of all direct children of ``parentPath`` in the
 * consolidated metadata, optionally filtered by node type.
 *
 * @param consolidated  The ``consolidated_metadata.metadata`` map.
 * @param parentPath    Relative path (no leading ``/``).
 *                      Pass an empty string to list top-level entries.
 * @param nodeType      Optional filter: ``"array"`` or ``"group"``.
 */
function childNames(
  consolidated: Record<string, ConsolidatedEntry>,
  parentPath: string,
  nodeType?: "array" | "group",
): string[] {
  const prefix = parentPath ? `${parentPath}/` : "";
  const depth = parentPath ? parentPath.split("/").length + 1 : 1;
  return Object.keys(consolidated)
    .filter(
      (p) =>
        p.startsWith(prefix) &&
        p.split("/").length === depth &&
        (nodeType === undefined || consolidated[p].node_type === nodeType),
    )
    .map((p) => p.split("/").at(-1)!);
}

// ---------------------------------------------------------------------------
// PosteriorSamples / PriorSamples
// ---------------------------------------------------------------------------

/**
 * Columnar posterior sample arrays for one analysis.
 *
 * Parameter names are available synchronously from the consolidated
 * metadata; the underlying float64 data is loaded lazily on demand.
 */
export class PosteriorSamples {
  /** @internal */
  constructor(
    private readonly rootLoc: zarrita.Location<Readable>,
    private readonly basePath: string,
    private readonly params: readonly string[],
  ) {}

  /** Names of all available posterior parameters. */
  parameterNames(): string[] {
    return [...this.params];
  }

  /**
   * Load the float64 data for a single parameter.
   *
   * @param param  Parameter name, e.g. ``"mass_1"``.
   * @throws If ``param`` is not in {@link parameterNames}.
   */
  async get(param: string): Promise<Float64Array> {
    if (!this.params.includes(param)) {
      throw new Error(`Parameter not found: ${param}`);
    }
    const arr = await zarrita.open(
      this.rootLoc.resolve(`${this.basePath}/${param}`),
      { kind: "array" },
    );
    const chunk = await zarrita.get(arr, null);
    // All Mitchell posterior arrays are float64 by schema convention.
    return chunk.data as Float64Array;
  }

  /**
   * Load all parameters into memory in parallel.
   *
   * @returns A mapping of parameter name → {@link Float64Array}.
   */
  async getAll(): Promise<Record<string, Float64Array>> {
    const entries = await Promise.all(
      this.params.map(async (p) => [p, await this.get(p)] as const),
    );
    return Object.fromEntries(entries);
  }
}

// ---------------------------------------------------------------------------
// Priors
// ---------------------------------------------------------------------------

/**
 * Prior information for one analysis: both sampled and analytic forms.
 */
export class Priors {
  /** @internal */
  constructor(
    private readonly rootLoc: zarrita.Location<Readable>,
    private readonly priorsPath: string,
    private readonly priorParamNames: readonly string[],
    /** Mapping of parameter name → prior description string. */
    readonly analytic: Readonly<Record<string, string>>,
  ) {}

  /** Whether the store contains sampled prior draws. */
  hasSamples(): boolean {
    return this.priorParamNames.length > 0;
  }

  /**
   * Sampled prior draws, or ``null`` if none were stored.
   *
   * The returned object exposes the same API as {@link PosteriorSamples}.
   */
  get samples(): PosteriorSamples | null {
    if (!this.hasSamples()) return null;
    return new PosteriorSamples(
      this.rootLoc,
      `${this.priorsPath}/samples`,
      this.priorParamNames,
    );
  }
}

// ---------------------------------------------------------------------------
// PSDs / CalibrationEnvelope
// ---------------------------------------------------------------------------

/** A 2-D float64 array with its shape metadata. */
export interface Array2D {
  /** Row-major (C-order) flattened data. */
  data: Float64Array;
  /** ``[rows, cols]``. */
  shape: [number, number];
}

/**
 * Power Spectral Density arrays for one analysis, keyed by detector name.
 *
 * Each PSD is a ``(N, 2)`` float64 array where the columns are
 * ``[frequency_Hz, ASD]``.
 */
export class PSDs {
  /** @internal */
  constructor(
    private readonly rootLoc: zarrita.Location<Readable>,
    private readonly psdsPath: string,
    private readonly detectors: readonly string[],
    private readonly consolidated: Record<string, ConsolidatedEntry>,
  ) {}

  /** Names of all detectors that have a PSD stored. */
  detectorNames(): string[] {
    return [...this.detectors];
  }

  /**
   * Load the PSD array for a detector.
   *
   * @param detector  Detector name, e.g. ``"H1"`` or ``"L1"``.
   * @throws If ``detector`` is not in {@link detectorNames}.
   */
  async get(detector: string): Promise<Array2D> {
    if (!this.detectors.includes(detector)) {
      throw new Error(`Detector not found: ${detector}`);
    }
    const arr = await zarrita.open(
      this.rootLoc.resolve(`${this.psdsPath}/${detector}`),
      { kind: "array" },
    );
    const chunk = await zarrita.get(arr, null);
    const shape = this.consolidated[`${this.psdsPath}/${detector}`].shape!;
    return { data: chunk.data as Float64Array, shape: shape as [number, number] };
  }
}

/**
 * Calibration envelope arrays for one analysis, keyed by detector name.
 *
 * Each envelope is a ``(300, 7)`` float64 array.
 */
export class CalibrationEnvelope {
  /** @internal */
  constructor(
    private readonly rootLoc: zarrita.Location<Readable>,
    private readonly calPath: string,
    private readonly detectors: readonly string[],
    private readonly consolidated: Record<string, ConsolidatedEntry>,
  ) {}

  /** Names of all detectors with a calibration envelope stored. */
  detectorNames(): string[] {
    return [...this.detectors];
  }

  /**
   * Load the calibration envelope array for a detector.
   *
   * @param detector  Detector name, e.g. ``"H1"``.
   * @throws If ``detector`` is not in {@link detectorNames}.
   */
  async get(detector: string): Promise<Array2D> {
    if (!this.detectors.includes(detector)) {
      throw new Error(`Detector not found: ${detector}`);
    }
    const arr = await zarrita.open(
      this.rootLoc.resolve(`${this.calPath}/${detector}`),
      { kind: "array" },
    );
    const chunk = await zarrita.get(arr, null);
    const shape = this.consolidated[`${this.calPath}/${detector}`].shape!;
    return { data: chunk.data as Float64Array, shape: shape as [number, number] };
  }
}

// ---------------------------------------------------------------------------
// Skymap
// ---------------------------------------------------------------------------

/**
 * HEALPix skymap for one analysis.
 */
export class Skymap {
  /** @internal */
  constructor(
    private readonly rootLoc: zarrita.Location<Readable>,
    private readonly skymapPath: string,
    /** Raw zarr attributes for the skymap group (HEALPix metadata). */
    readonly metadata: Record<string, unknown>,
  ) {}

  /** Load the HEALPix pixel data as a flat float64 array. */
  async getData(): Promise<Float64Array> {
    const arr = await zarrita.open(
      this.rootLoc.resolve(`${this.skymapPath}/data`),
      { kind: "array" },
    );
    const chunk = await zarrita.get(arr, null);
    return chunk.data as Float64Array;
  }
}

// ---------------------------------------------------------------------------
// Analysis
// ---------------------------------------------------------------------------

/**
 * A single parameter-estimation analysis (one waveform approximant run)
 * within an event.
 *
 * All data-product accessors (``posterior``, ``psds``, …) are synchronous
 * and return ``null`` when a product was not stored for this analysis.
 * Actual array data is loaded asynchronously via the returned objects.
 */
export class Analysis {
  /** @internal */
  constructor(
    /** Analysis name, e.g. ``"IMRPhenomXPHM"``. */
    readonly name: string,
    /**
     * Analysis metadata attributes from the zarr group.
     *
     * Includes waveform approximant, frequency settings, and other
     * bilby meta_data fields promoted at import time.
     */
    readonly attrs: AnalysisAttrs,
    /** Posterior sample arrays, or ``null`` if not stored. */
    readonly posterior: PosteriorSamples | null,
    /** Prior information, or ``null`` if not stored. */
    readonly priors: Priors | null,
    /** PSD arrays, or ``null`` if not stored. */
    readonly psds: PSDs | null,
    /** Calibration envelope arrays, or ``null`` if not stored. */
    readonly calibrationEnvelope: CalibrationEnvelope | null,
    /** HEALPix skymap, or ``null`` if not stored. */
    readonly skymap: Skymap | null,
    /** Config file key-value pairs as a plain object. */
    readonly config: Record<string, unknown>,
  ) {}

  toString(): string {
    return `Analysis("${this.name}")`;
  }
}

// ---------------------------------------------------------------------------
// Analysis factory
// ---------------------------------------------------------------------------

/** Build an Analysis purely from consolidated metadata (no zarr I/O). */
function buildAnalysis(
  rootLoc: zarrita.Location<Readable>,
  consolidated: Record<string, ConsolidatedEntry>,
  analysisPath: string,
): Analysis {
  const name = analysisPath.split("/").at(-1)!;
  const attrs = (consolidated[analysisPath]?.attributes ?? {}) as AnalysisAttrs;

  // Posterior samples
  const postSamplesPath = `${analysisPath}/posterior/samples`;
  const postParams = childNames(consolidated, postSamplesPath, "array");
  const posterior =
    postParams.length > 0
      ? new PosteriorSamples(rootLoc, postSamplesPath, postParams)
      : null;

  // Prior samples + analytic priors
  const priorsPath = `${analysisPath}/priors`;
  let priors: Priors | null = null;
  if (priorsPath in consolidated) {
    const priorSamplesPath = `${priorsPath}/samples`;
    const priorParams = childNames(consolidated, priorSamplesPath, "array");
    const analyticPath = `${priorsPath}/analytic`;
    const analyticAttrs =
      analyticPath in consolidated
        ? (consolidated[analyticPath].attributes as Record<string, string>)
        : {};
    priors = new Priors(rootLoc, priorsPath, priorParams, analyticAttrs);
  }

  // PSDs
  const psdsPath = `${analysisPath}/psds`;
  const psdDetectors = childNames(consolidated, psdsPath, "array");
  const psds =
    psdDetectors.length > 0
      ? new PSDs(rootLoc, psdsPath, psdDetectors, consolidated)
      : null;

  // Calibration envelope
  const calPath = `${analysisPath}/calibration_envelope`;
  const calDetectors = childNames(consolidated, calPath, "array");
  const calibrationEnvelope =
    calDetectors.length > 0
      ? new CalibrationEnvelope(rootLoc, calPath, calDetectors, consolidated)
      : null;

  // Skymap
  const skymapPath = `${analysisPath}/skymap`;
  const skymapDataPath = `${skymapPath}/data`;
  const skymap =
    skymapDataPath in consolidated
      ? new Skymap(
          rootLoc,
          skymapPath,
          consolidated[skymapPath]?.attributes ?? {},
        )
      : null;

  // Config file
  const configPath = `${analysisPath}/config_file`;
  const config =
    configPath in consolidated ? consolidated[configPath].attributes : {};

  return new Analysis(
    name,
    attrs,
    posterior,
    priors,
    psds,
    calibrationEnvelope,
    skymap,
    config,
  );
}

// ---------------------------------------------------------------------------
// Event
// ---------------------------------------------------------------------------

/**
 * A gravitational-wave event containing one or more analyses.
 */
export class Event {
  /** @internal */
  constructor(
    private readonly rootLoc: zarrita.Location<Readable>,
    private readonly consolidated: Record<string, ConsolidatedEntry>,
    /** Event name, e.g. ``"GW150914_095045"``. */
    readonly name: string,
  ) {}

  /** Names of all analyses stored for this event. */
  analysisNames(): string[] {
    return childNames(this.consolidated, `events/${this.name}`);
  }

  /** Total number of analyses stored for this event. */
  size(): number {
    return this.analysisNames().length;
  }

  /**
   * Retrieve an analysis by name.
   *
   * @param analysisName  e.g. ``"IMRPhenomXPHM"``.
   * @throws If the analysis is not found.
   */
  async get(analysisName: string): Promise<Analysis> {
    const path = `events/${this.name}/${analysisName}`;
    if (!(path in this.consolidated)) {
      throw new Error(`Analysis not found: ${analysisName}`);
    }
    return buildAnalysis(this.rootLoc, this.consolidated, path);
  }

  /** Async-iterate over all analyses for this event. */
  async *analyses(): AsyncGenerator<Analysis> {
    for (const name of this.analysisNames()) {
      yield buildAnalysis(
        this.rootLoc,
        this.consolidated,
        `events/${this.name}/${name}`,
      );
    }
  }

  toString(): string {
    return `Event("${this.name}", analyses=[${this.analysisNames().join(", ")}])`;
  }
}

// ---------------------------------------------------------------------------
// MitchellStore
// ---------------------------------------------------------------------------

/**
 * A zarr-backed store of gravitational-wave analysis data products.
 *
 * Open an existing store with {@link MitchellStore.open}, passing any
 * zarrita-compatible store object.  For local filesystem access from
 * Node.js use {@link FileSystemStore} from ``@mitchell/javascript/node``;
 * for HTTP/cloud access use zarrita's built-in ``FetchStore``.
 *
 * Discovery (listing events and analyses) is performed synchronously
 * using consolidated metadata loaded at open time.
 *
 * @example Open a local store (Node.js)
 * ```typescript
 * import { FileSystemStore } from "@mitchell/javascript/node";
 * import { MitchellStore } from "@mitchell/javascript";
 *
 * const ms = await MitchellStore.open(new FileSystemStore("events.zarr"));
 * console.log(ms.eventNames()); // ["GW150914_095045", ...]
 * ```
 *
 * @example Open a remote store (browser)
 * ```typescript
 * import * as zarrita from "zarrita";
 * import { MitchellStore } from "@mitchell/javascript";
 *
 * const ms = await MitchellStore.open(
 *   new zarrita.FetchStore("https://my-bucket.s3.amazonaws.com/events.zarr")
 * );
 * ```
 */
export class MitchellStore {
  /** The Mitchell schema version of this store. */
  readonly schemaVersion: string;

  private constructor(
    private readonly rootLoc: zarrita.Location<Readable>,
    private readonly consolidated: Record<string, ConsolidatedEntry>,
    schemaVersion: string,
  ) {
    this.schemaVersion = schemaVersion;
  }

  /**
   * Open a Mitchell store from any zarrita-compatible store object.
   *
   * Reads and parses the root ``zarr.json`` (including consolidated
   * metadata) in a single request.
   *
   * @param store  Any {@link Readable} store.  Accepts
   *               {@link FileSystemStore} (Node.js), ``zarrita.FetchStore``
   *               (browser/HTTP), or an in-memory
   *               ``Map<string, Uint8Array>``.
   * @throws If ``/zarr.json`` is absent, the node is not a zarr v3 group,
   *         or ``mitchell_schema_version`` is missing from root attributes.
   */
  static async open(store: Readable): Promise<MitchellStore> {
    const rootMeta = await readRootJson(store);

    if (rootMeta.zarr_format !== 3 || rootMeta.node_type !== "group") {
      throw new Error(
        "Not a valid zarr v3 group at the store root",
      );
    }

    const version = rootMeta.attributes?.mitchell_schema_version;
    if (!version) {
      throw new Error(
        "Not a Mitchell store: mitchell_schema_version attribute not found in root",
      );
    }

    const consolidated = rootMeta.consolidated_metadata?.metadata ?? {};
    const rootLoc = zarrita.root(store);
    return new MitchellStore(rootLoc, consolidated, version);
  }

  /** Names of all events in the store (from consolidated metadata). */
  eventNames(): string[] {
    return childNames(this.consolidated, "events");
  }

  /** Total number of events in the store. */
  size(): number {
    return this.eventNames().length;
  }

  /**
   * Retrieve an event by name.
   *
   * @param eventName  e.g. ``"GW150914_095045"``.
   * @throws If the event is not found.
   */
  async get(eventName: string): Promise<Event> {
    if (!this.eventNames().includes(eventName)) {
      throw new Error(`Event not found: ${eventName}`);
    }
    return new Event(this.rootLoc, this.consolidated, eventName);
  }

  /** Async-iterate over all events in the store. */
  async *events(): AsyncGenerator<Event> {
    for (const name of this.eventNames()) {
      yield new Event(this.rootLoc, this.consolidated, name);
    }
  }

  toString(): string {
    const n = this.size();
    return `MitchellStore(${n} event${n !== 1 ? "s" : ""})`;
  }
}
